File: 100-doors/csharp/src/HundredDoors/HundredDoors.cs
namespace HundredDoors;

public static class HundredDoors
{
    public static IReadOnlyList<int> OpenDoors(int numDoors)
    {
        var isOpen = new bool[numDoors + 1];
        for (var pass = 1; pass <= numDoors; pass++)
        {
            for (var door = pass; door <= numDoors; door += pass)
            {
                isOpen[door] = !isOpen[door];
            }
        }

        var open = new List<int>();
        for (var door = 1; door <= numDoors; door++)
        {
            if (isOpen[door]) open.Add(door);
        }
        return open;
    }
}

File: 100-doors/typescript/src/hundredDoors.ts
export function openDoors(numDoors: number): readonly number[] {
  const isOpen = new Array<boolean>(numDoors + 1).fill(false);
  for (let pass = 1; pass <= numDoors; pass++) {
    for (let door = pass; door <= numDoors; door += pass) {
      isOpen[door] = !isOpen[door];
    }
  }

  const open: number[] = [];
  for (let door = 1; door <= numDoors; door++) {
    if (isOpen[door]) open.push(door);
  }
  return open;
}
